### ISSUE TITLE
autosuggest

### ISSUE THREAD

**Issue by 257** [fanglingsu/vimb#180]
...much like google-chrome's add bar. that is put the first completion candidate in front of the cursor in white-on-blue (customisable of course) font.
i know that qt supports this, dunno about gtk.

important note:
most implementations i have seen get this wrong (auto-fu for zsh is one); backspace is to wipe the suggestion _first_ before starting to delete typed chars by user. so say bar is type and baz is suggested and bar_baz_ is shown (\* marker is to say white-on-blue or bold or whatever). now user hits backspace, this will wipe baz and leave bar. then user hits backspace again, this deletes 'r' and so on.

**Comment by fanglingsu** [comment]
I don't like the concept of autosuggest. This is bloated and wasts cpu cycles without any real benefit. Why should we suggest something when the user is still typing and don't want's to the  distracted by a suggestion list. And as I can remember the timer when I used chromium, the suggested urls where rarely those I expected to appear.
Vimb also tries to mimik the behaviour and feel of vim, so a autosuggest for the inputbox would be unexpected.

On the other side there might be some other users like you that want to have such a behaviour. So I mark this issue as request for comment to see what's the opinion of the other users.

**Comment by iamleot** [comment]
Hello Daniel and the entire vimb community,

Daniel Carl writes:
> I don't like the concept of autosuggest. This is bloated and wasts
> cpu cycles without any real benefit. Why should we suggest something
> when the user is still typing and don't want's to the  distracted by
> a suggestion list.
> [...]
> On the other side there might be some other users like you that want
> to have such a behaviour. So I mark this issue as request for comment
> to see what's the opinion of the other users.
+1, I completely agree with you regarding the bloat and that it will not
bring any real benefit. We have "shortcut-add" that does a great job.


Ciao,
L.

**Comment by foo-bar-baz** [comment]
On Feb 17  14:18 -0800, Daniel Carl wrote:
> I don't like the concept of autosuggest. This is bloated and wasts cpu
> cycles without any real benefit. Why should we suggest something when the
> user is still typing and don't want's to the  distracted by a suggestion
> list. 

or perhaps enlightened! someone starting to use your program today with no
notion of commands available to him has to RTFM from a-to-z to know what's
there. well it turns out that people don't like to RTFM like that and it's
the right attitude since manuals are for references (like dictionaries).
this goes back to the idea of cli loosing to gui because gui gives user 
options they can choose from, albeit limited number of choices compared to
what's available. sort of a dumbed down version of things. but it baffles
me why cli designers rarely do the same. they can do much better and it's
actually is a must in cli since _none_ can remember all the commands.  and
it's not just commands, combination of arguments you can pass to them too.
and they tend to get long in a hurry. say you have a command :foo and it
takes bar and baz as arguments and user has issued :foo bar and :foo baz
and :foo bar baz in the past. now if he want to issue the last one again to
you think it's fun to type the whole thing all over again? of course not,
he wants to type 'r b' and get it done with.
a clunky implementation of what i'm talking about is auto-fu for zsh. with
all it's short comings it's cli on steroid.

> And as I can remember the timer when I used chromium, the suggested
> urls where rarely those I expected to appear.

i mean this issue to be more like omni bar in other browser but bar the
search engine suggestion. i have filed another ticket for that. they are to
be merge at some point of course.
i will elaborate on this there.

> Vimb also tries to mimik the behaviour and feel of vim, so a autosuggest
> for the inputbox would be unexpected.

that's taking the idea of and editor in a browser a bit literally. and you
wouldn't be the first. vimperator itself (first to implement the idea i
think) gets it wrong. a browser is more like a pager (since you can't edit
much) rather than an editor. so i would say ``less'' would have been a
better analogy...

but even if we take the idea of mimiking literally as you suggest, it's not
true that vim doesn't do autosuggestion. i'm writing this on vim and
youcompleme keeps popping up suggestions left and right. (doesn't put it in
front of my cursor though which is a shame)

> On the other side there might be some other users like you that want to
> have such a behaviour. So I mark this issue as request for comment to see
> what's the opinion of the other users.

can be easily hidden behind a :set setting. not everyone needs to know
about it.

**Comment by fanglingsu** [comment]
As I understand you right you are interested in a better completion/suggestion for the things in the inputbox (cli of vimb). Right? And you want to easily find previous entered commands with parameters? This is still available by starting command mode via `:` and type some text and hit `<Tab>`. But the completions are filtered by prefix match and you can't type ahead to near the suggestions down.

At the time the completions list available commands or in case the command is known possible next options for it. And this is good in case you don't have a large history of old commands. If I understood you right, you mean that there should be an additional completion that uses previous typed commands too. I don't know auto-fu for zsh so it's not easy for me to imagine how this should behave and by which algorithm the suggestions are done. OK if you type 'r b' it's clear that this isn't a known command, but what for ':set '? This would lead to a completion list with all posible settings plus all set commands ever run from vimb.

**Comment by foo-bar-baz** [comment]
On Feb 18  14:22 -0800, Daniel Carl wrote:
> As I understand you right you are interested in a better
> completion/suggestion for the things in the inputbox (cli of vimb).
> Right?

correct.

> And you want to easily find previous entered commands with
> parameters? This is still available by starting command mode via `:` and
> type some text and hit `<Tab>`. 

this is, so far, like vimperator. i think i mentioned in my very first post
that the idea is to get to chrome's addr bar. or fsf, see the pic in the
next post.

> But the completions are filtered by prefix match and you can't type ahead
> to near the suggestions down.
> 
> At the time the completions list available commands or in case the
> command is known possible next options for it. And this is good in case
> you don't have a large history of old commands. If I understood you
> right, you mean that there should be an additional completion that uses
> previous typed commands too.

yes. lets call it 'history'. this includes but it's not limited to what is
traditionally called history in browser; arguments to :open in vimb. for
example it also includes arguments passed to :set as well and so on. more
like history in shells if you like.

> I don't know auto-fu for zsh so it's not
> easy for me to imagine how this should behave and by which algorithm the
> suggestions are done. OK if you type 'r b' it's clear that this isn't a
> known command, but what for ':set '? This would lead to a completion list
> with all posible settings plus all set commands ever run from vimb.

you pretty much got it. plus history should almost always come first.
(coming-first part can be made configurable of course, i'm not a dictator).

the ideal situation though is when we can let go of the old notion of
mechanically typing :set first and then it's argument(s) and then hitting
Enter... just type the as much as you remember, and please try to be clever
and type the part that's unique, and we should be able to find you what
you're looking for. even if we miss user still would be presented with good
enough template to edit and get to where he wants to.

**Comment by foo-bar-baz** [comment]
https://camo.githubusercontent.com/0b07def9e05309281212369b118fcf9b9fc7948e/68747470733a2f2f7261772e6769746875622e636f6d2f6a756e6567756e6e2f692f6d61737465722f667a662e676966

![68747470733a2f2f7261772e6769746875622e636f6d2f6a756e6567756e6e2f692f6d61737465722f667a662e676966](https://cloud.githubusercontent.com/assets/1268392/6261080/5afcc006-b7b7-11e4-8f77-ae5314ceecbb.gif)

**Comment by sdothum** [comment]
My lowly 2cents on this issue, without any data to back it up, is that people attracted to vimb probably came to it for its leanness -- both in footprint and UI. I don't see very well the application with regards to vimb settings (again, because vimb users will cull through the docs to fine tune it just the way they want). For history and bookmarks, I like typing a few (partial) keywords and press TAB. Granted, sometimes that yields a bigger set than I expected -- some way to go back into command line input mode might be nice to further refine the search. Because I keep such a large history, I set up vimb to use a pruned version for performance issues and use dmenu for the full history -- with dmenu I have real time display and filtering of the history file as I type: quite effective and I can use that mechanism on recent history as well, though, it is not as seamless as the vimb command line.
